extensions/metrics/src/interceptors/metrics.interceptor.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/extension-metrics
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  asGlobalInterceptor,
+  bind,
+  BindingScope,
+  Interceptor,
+  InvocationContext,
+  Provider,
+  ValueOrPromise,
+} from '@loopback/context';
+import {Counter, Gauge, Histogram, Summary} from 'prom-client';
+
+/**
+ * This interceptor captures metrics for method invocations,
+ * excluding sequence actions and middleware executed before
+ * a method is invoked. Please collect metrics at other places
+ * if you want to cover more than just method invocations.
+ */
+@bind(asGlobalInterceptor('metrics'), {scope: BindingScope.SINGLETON})
+export class MetricsInterceptor implements Provider<Interceptor> {
+  private gauge = new Gauge({
+    name: 'loopback_invocation_duration_seconds',
+    help: 'method invocation',
+    labelNames: ['targetName'],
+  });
+
+  private histogram = new Histogram({
+    name: 'loopback_invocation_duration_histogram',
+    help: 'method invocation histogram',
+    labelNames: ['targetName'],
+  });
+
+  private counter = new Counter({
+    name: 'loopback_invocation_total',
+    help: 'method invocation counts',
+    labelNames: ['targetName'],
+  });
+
+  private summary = new Summary({
+    name: 'loopback_invocation_duration_summary',
+    help: 'method invocation summary',
+    labelNames: ['targetName'],
+  });
+
+  constructor() {}
+
+  value() {
+    return this.intercept.bind(this);
+  }
+
+  async intercept<T>(
+    invocationCtx: InvocationContext,
+    next: () => ValueOrPromise<T>,
+  ) {
+    const endGauge = this.gauge.startTimer({
+      targetName: invocationCtx.targetName,
+    });
+    const endHistogram = this.histogram.startTimer({
+      targetName: invocationCtx.targetName,
+    });
+    const endSummary = this.summary.startTimer({
+      targetName: invocationCtx.targetName,
+    });
+    try {
+      this.counter.inc();
```

Comment:
A question for the counter: I run the demo and noticed the method invocation # is 105:
```
# HELP loopback_invocation_total method invocation counts
# TYPE loopback_invocation_total counter
loopback_invocation_total 105
```
The demo app doesn't have any controller/endpoints, so I am wondering... what are the methods being invoked?

Reply:
Good question. We have a [built-in controller](https://github.com/strongloop/loopback-next/blob/prometheus/extensions/metrics/src/controllers/metrics.controller.ts) in `extensions/metrics`.

So whenever the `/metrics` is scraped, our metrics interceptor is triggered.

Reply:
ah, that makes sense 👍